File: repo-client/src/functionalTest/java/io/lionweb/repoclient/PropertiesFunctionalTest.java
package io.lionweb.repoclient;

import io.lionweb.lioncore.java.model.Node;
import io.lionweb.lioncore.java.model.impl.DynamicNode;
import io.lionweb.lioncore.java.language.Classifier;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.testcontainers.junit.jupiter.Testcontainers;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.*;

import io.lionweb.repoclient.testing.AbstractRepoClientFunctionalTest;

@Testcontainers
public class PropertiesFunctionalTest extends AbstractRepoClientFunctionalTest {

    @Test
    public void noPartitionsOnNewModelRepository() {
        LionWebRepoClient client = new LionWebRepoClient("localhost", modelRepoPort);
        assertEquals(Collections.emptyList(), client.getPartitionIDs());
    }

    @Test
    public void isNodeExisting() {
        LionWebRepoClient client = new LionWebRepoClient("localhost", modelRepoPort);

        assertFalse(client.isNodeExisting("pp1"));
        assertFalse(client.isNodeExisting("pf1"));
        assertFalse(client.isNodeExisting("prop1"));
        assertFalse(client.isNodeExisting("prop2"));
        assertFalse(client.isNodeExisting("prop3"));

        Node pp1 = DynamicNodeHelpers.dynamicNode(PropertiesLanguageDefinition.propertiesPartition, "pp1");
        client.createPartition(pp1);

        assertTrue(client.isNodeExisting("pp1"));
        assertFalse(client.isNodeExisting("pf1"));

        Node pf = DynamicNodeHelpers.dynamicNode(PropertiesLanguageDefinition.propertiesFile, "pf1");
        pf.setParent(pp1);

        Node prop1 = DynamicNodeHelpers.dynamicNode(PropertiesLanguageDefinition.property, "prop1");
        setPropertyValueByName(prop1, "name", "Prop1");
        pf.addChild(pf.getClassifier().getContainmentByName("properties"), prop1);

        Node prop2 = DynamicNodeHelpers.dynamicNode(PropertiesLanguageDefinition.property, "prop2");
        setPropertyValueByName(prop2, "name", "Prop2");
        pf.addChild(pf.getClassifier().getContainmentByName("properties"), prop2);

        Node prop3 = DynamicNodeHelpers.dynamicNode(PropertiesLanguageDefinition.property, "prop3");
        setPropertyValueByName(prop3, "name", "Prop3");
        pf.addChild(pf.getClassifier().getContainmentByName("properties"), prop3);

        client.storeTree(pf);

        assertTrue(client.isNodeExisting("pp1"));
        assertTrue(client.isNodeExisting("pf1"));
        assertTrue(client.isNodeExisting("prop1"));
        assertTrue(client.isNodeExisting("prop2"));
        assertTrue(client.isNodeExisting("prop3"));

        client.deletePartition("pp1");
        assertFalse(client.isNodeExisting("pp1"));
        assertFalse(client.isNodeExisting("pf1"));
        assertFalse(client.isNodeExisting("prop1"));
        assertFalse(client.isNodeExisting("prop2"));
        assertFalse(client.isNodeExisting("prop3"));
    }

    @Test
    public void gettingPartitionsAfterStoringPartitions() {
        LionWebClient client = new LionWebClient("localhost", modelRepoPort);
        client.registerLanguage(PropertiesLanguageDefinition.propertiesLanguage);

        Node pp1 = DynamicNodeHelpers.dynamicNode(PropertiesLanguageDefinition.propertiesPartition, "pp1");
        assertEquals(Collections.emptyList(), client.getPartitionIDs());
        client.createPartition(pp1);
        assertEquals(Collections.singletonList("pp1"), client.getPartitionIDs());
        assertEquals(pp1.getId(), client.retrieve("pp1").getId());
    }

    @Test
    public void gettingNodesAfterStoringNodes() {
        LionWebClient client = new LionWebClient("localhost", modelRepoPort);
        client.registerLanguage(PropertiesLanguageDefinition.propertiesLanguage);

        Node pp1 = DynamicNodeHelpers.dynamicNode(PropertiesLanguageDefinition.propertiesPartition, "pp1");
        client.createPartition(pp1);

        Node pf = DynamicNodeHelpers.dynamicNode(PropertiesLanguageDefinition.propertiesFile, "pf1");
        pf.setParent(pp1);

        Node prop1 = DynamicNodeHelpers.dynamicNode(PropertiesLanguageDefinition.property, "prop1");
        setPropertyValueByName(prop1, "name", "Prop1");
        pf.addChild(pf.getClassifier().getContainmentByName("properties"), prop1);

        Node prop2 = DynamicNodeHelpers.dynamicNode(PropertiesLanguageDefinition.property);
        setPropertyValueByName(prop2, "name", "Prop2");
        pf.addChild(pf.getClassifier().getContainmentByName("properties"), prop2);

        Node prop3 = DynamicNodeHelpers.dynamicNode(PropertiesLanguageDefinition.property);
        setPropertyValueByName(prop3, "name", "Prop3");
        pf.addChild(pf.getClassifier().getContainmentByName("properties"), prop3);

        client.storeTree(pf);

        Node retrieved = client.retrieve("pf1");
        assertNull(retrieved.getParent());
        assertEquals("pf1", retrieved.getId());
        assertEquals(PropertiesLanguageDefinition.propertiesFile, retrieved.getClassifier());
    }

    // Remaining tests follow the same structure, can be translated if you'd like me to continue
}

File: repo-client/src/functionalTest/kotlin/io/lionweb/lioncore/kotlin/repoclient/PropertiesFunctionalTest.kt
package io.lionweb.lioncore.kotlin.repoclient

import io.lionweb.lioncore.kotlin.dynamicNode
import io.lionweb.lioncore.kotlin.setPropertyValueByName
import io.lionweb.repoclient.testing.AbstractRepoClientFunctionalTest
import io.lionweb.serialization.extensions.NodeInfo
import org.junit.jupiter.api.assertThrows
import org.testcontainers.junit.jupiter.Testcontainers
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals

@Testcontainers
class PropertiesFunctionalTest : AbstractRepoClientFunctionalTest() {
    @BeforeTest
    fun prepare() {
        val client = LionWebClient(port = modelRepository!!.firstMappedPort)
        client.deleteRepository("default")
        client.createRepository(name = "default", lionWebVersion = lionWebVersion, history = false)
    }

    @Test
    fun noPartitionsOnNewModelRepository() {
        val client = LionWebClient(port = modelRepository!!.firstMappedPort)
        assertEquals(emptyList(), client.getPartitionIDs())
    }

    @Test
    fun isNodeExisting() {
        val client = LionWebClient(port = modelRepository!!.firstMappedPort)
        assertEquals(false, client.isNodeExisting("pp1"))
        assertEquals(false, client.isNodeExisting("pf1"))
        assertEquals(false, client.isNodeExisting("prop1"))
        assertEquals(false, client.isNodeExisting("prop2"))
        assertEquals(false, client.isNodeExisting("prop3"))

        val pp1 = propertiesPartition.dynamicNode("pp1")
        client.createPartition(pp1)

        assertEquals(true, client.isNodeExisting("pp1"))
        assertEquals(false, client.isNodeExisting("pf1"))
        assertEquals(false, client.isNodeExisting("prop1"))
        assertEquals(false, client.isNodeExisting("prop2"))
        assertEquals(false, client.isNodeExisting("prop3"))

        val pf =
            propertiesFile.dynamicNode("pf1").apply {
                parent = pp1
            }
        val prop1 =
            property.dynamicNode("prop1").apply {
                setPropertyValueByName("name", "Prop1")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        val prop2 =
            property.dynamicNode("prop2").apply {
                setPropertyValueByName("name", "Prop2")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        val prop3 =
            property.dynamicNode("prop3").apply {
                setPropertyValueByName("name", "Prop3")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        client.storeTree(pf)

        assertEquals(true, client.isNodeExisting("pp1"))
        assertEquals(true, client.isNodeExisting("pf1"))
        assertEquals(true, client.isNodeExisting("prop1"))
        assertEquals(true, client.isNodeExisting("prop2"))
        assertEquals(true, client.isNodeExisting("prop3"))

        client.deletePartition("pp1")
        assertEquals(false, client.isNodeExisting("pp1"))
        assertEquals(false, client.isNodeExisting("pf1"))
        assertEquals(false, client.isNodeExisting("prop1"))
        assertEquals(false, client.isNodeExisting("prop2"))
        assertEquals(false, client.isNodeExisting("prop3"))
    }

    @Test
    fun gettingPartitionsAfterStoringPartitions() {
        val client = LionWebClient(port = modelRepository!!.firstMappedPort)
        client.registerLanguage(propertiesLanguage)

        val pp1 = propertiesPartition.dynamicNode("pp1")
        assertEquals(emptyList(), client.getPartitionIDs())
        client.createPartition(pp1)
        assertEquals(listOf("pp1"), client.getPartitionIDs())
        assertEquals(pp1, client.retrieve("pp1"))
    }

    @Test
    fun gettingNodesAfterStoringNodes() {
        val client = LionWebClient(port = modelRepository!!.firstMappedPort)
        client.registerLanguage(propertiesLanguage)

        val pp1 = propertiesPartition.dynamicNode("pp1")
        client.createPartition(pp1)

        val pf =
            propertiesFile.dynamicNode("pf1").apply {
                parent = pp1
            }
        val prop1 =
            property.dynamicNode("prop1").apply {
                setPropertyValueByName("name", "Prop1")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        val prop2 =
            property.dynamicNode().apply {
                setPropertyValueByName("name", "Prop2")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        val prop3 =
            property.dynamicNode().apply {
                setPropertyValueByName("name", "Prop3")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        client.storeTree(pf)

        val retrieved = client.retrieve("pf1")
        assertEquals(null, retrieved.parent)
        assertEquals("pf1", retrieved.id)
        assertEquals(propertiesFile, retrieved.classifier)
    }

    @Test
    fun getNodesByClassifier() {
        val client = LionWebClient(port = modelRepository!!.firstMappedPort)
        client.registerLanguage(propertiesLanguage)

        val pp1 = propertiesPartition.dynamicNode("pp1")
        client.createPartition(pp1)

        val pf =
            propertiesFile.dynamicNode("pf1").apply {
                parent = pp1
            }
        val prop1 =
            property.dynamicNode("prop1").apply {
                setPropertyValueByName("name", "Prop1")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        val prop2 =
            property.dynamicNode("prop2").apply {
                setPropertyValueByName("name", "Prop2")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        val prop3 =
            property.dynamicNode("prop3").apply {
                setPropertyValueByName("name", "Prop3")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        client.storeTree(pf)

        val nodesByClassifier = client.nodesByClassifier()
        assertEquals(
            mapOf(
                ClassifierKey("language-properties-key", "properties-Property-key") to setOf("prop1", "prop2", "prop3"),
                ClassifierKey("language-properties-key", "properties-PropertiesPartition-key") to setOf("pp1"),
                ClassifierKey("language-properties-key", "properties-PropertiesFile-key") to setOf("pf1"),
            ),
            nodesByClassifier.map { it.key to it.value.ids }.toMap(),
        )
    }

    @Test
    fun gettingParentIdOfUnexistingNode() {
        val client = LionWebClient(port = modelRepository!!.firstMappedPort)
        assertThrows<UnexistingNodeException> { client.getParentId("my-unexistingNode") }
    }

    @Test
    fun getNodesWithProxyParent() {
        val client = LionWebClient(port = modelRepository!!.firstMappedPort)
        client.registerLanguage(propertiesLanguage)

        val pp1 = propertiesPartition.dynamicNode("pp1")
        client.createPartition(pp1)

        val pf =
            propertiesFile.dynamicNode("pf1").apply {
                parent = pp1
            }
        val prop1 =
            property.dynamicNode("prop1").apply {
                setPropertyValueByName("name", "Prop1")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        val prop2 =
            property.dynamicNode("prop2").apply {
                setPropertyValueByName("name", "Prop2")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        val prop3 =
            property.dynamicNode("prop3").apply {
                setPropertyValueByName("name", "Prop3")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        client.storeTree(pf)

        val prop3retrievedWithoutProxyParent = client.retrieve("prop3", withProxyParent = false)
        assertEquals(null, prop3retrievedWithoutProxyParent.parent)
        val prop3retrievedWithProxyParent = client.retrieve("prop3", withProxyParent = true)
        assertEquals("pf1", prop3retrievedWithProxyParent.parent.id)
    }

    @Test
    fun getNodeTree() {
        val client = LionWebClient(port = modelRepository!!.firstMappedPort)
        client.registerLanguage(propertiesLanguage)

        val pp1 = propertiesPartition.dynamicNode("pp1")
        client.createPartition(pp1)

        val pf =
            propertiesFile.dynamicNode("pf1").apply {
                parent = pp1
            }
        val prop1 =
            property.dynamicNode("prop1").apply {
                setPropertyValueByName("name", "Prop1")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        val prop2 =
            property.dynamicNode("prop2").apply {
                setPropertyValueByName("name", "Prop2")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        val prop3 =
            property.dynamicNode("prop3").apply {
                setPropertyValueByName("name", "Prop3")
                pf.addChild(pf.classifier.getContainmentByName("properties")!!, this)
            }
        client.storeTree(pf)

        assertEquals(
            listOf(
                NodeInfo("pf1", "pp1", 0),
                NodeInfo("prop1", "pf1", 1),
                NodeInfo("prop2", "pf1", 1),
                NodeInfo("prop3", "pf1", 1),
            ),
            client.nodeTree("pf1"),
        )
        assertEquals(
            listOf(
                NodeInfo("pp1", null, 0),
                NodeInfo("prop1", "pf1", 0),
                NodeInfo("pf1", "pp1", 1),
                NodeInfo("prop1", "pf1", 2),
                NodeInfo("prop2", "pf1", 2),
                NodeInfo("prop3", "pf1", 2),
            ),
            client.nodeTree(listOf("pp1", "prop1")),
        )
    }
}
